### PR TITLE
Interface instance typechecking (haskell)

### PR DESCRIPTION
**Pull request by akrmn**

### CODE REVIEW

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 1]
```diff
@@ -938,34 +955,53 @@ checkTemplate m t@(Template _loc tpl param precond signatories observers text ch
...
-checkIfaceImplementation :: MonadGamma m => Template -> TemplateImplements -> m ()
-checkIfaceImplementation Template{tplTypeCon, tplImplements} TemplateImplements{..} = do
-  DefInterface {intRequires, intMethods, intView} <- inWorld $ lookupInterface tpiInterface
-  let InterfaceInstanceBody {iiMethods, iiView} = tpiBody
+checkInterfaceInstance :: MonadGamma m =>
+     ExprVarName
+  -> InterfaceInstanceKey
+  -> InterfaceInstanceBody
+  -> m ()
+checkInterfaceInstance tmplParam iiKey iiBody = do
```
Comment (akrmn) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs:972]: This is the main change in this PR. The approach is the same as in the corresponding scala function; `checkInterfaceInstance` doesn't care where the interface instance was defined, rather it's up to the caller to provide the right arguments and set up the context for error messages

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs [hunk 2]
```diff
@@ -41,6 +42,13 @@ data TemplatePart
   | TPAgreement
   | TPKey
   | TPChoice TemplateChoice
+  | TPInterfaceInstance InterfaceInstanceKey
```
Comment (akrmn) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs:45]: I extended `TemplatePart` with a new constructor `TPInterfaceInstance`, mirroring `TPChoice`.

However, even this falls short of the context information I'd like to have: if there's a type error in the body of a method `foo` in `interface instance Bar for Baz` in `template Baz`, the context doesn't mention the method. We could add another constructor to `TemplatePart`, 

```haskell
TPInterfaceInstanceMethod :: InterfaceInstanceKey -> MethodName -> TemplatePart
```

but then `checkInterfaceInstance` would need to construct a `Context` for each method, so it would either need to take a `Context`-making function or the `Template` itself. Even then, we would need to duplicate this for `InterfacePart`.

**tl;dr:** I think a better solution (simpler code and clearer error messages) would be to flatten the `Context` data type (and `TemplatePart`) and change [`Gamma`](https://github.com/digital-asset/daml/blob/main/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Env.hs#L36-L37) to contain a (perhaps non-empty) stack of `Context`s

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs [hunk 3]
```diff
@@ -158,7 +165,7 @@ contextLocation = \case
   ContextTemplate _ t _  -> tplLocation t
```
Comment (akrmn) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs:165]: this made me sad: even if the `TemplatePart` has a location somewhere, the error message only shows the location of the template body

File: compiler/daml-lf-ast/src/DA/Daml/LF/TemplateOrInterface.hs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+-- Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module DA.Daml.LF.TemplateOrInterface
```
Comment (remyhaemmerle-da) [inline compiler/daml-lf-ast/src/DA/Daml/LF/TemplateOrInterface.hs:4]: I do not really know Haskell convention.  But it is really worthwhile to create a new file for that ? Putting it in `Ast` or `Util` would not be good enought ?

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs (hunk 2 above)
Comment (remyhaemmerle-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs:45]: I let you judge for that, as long you make it in another PR.

**Review by remyhaemmerle-da — APPROVED** [review]
Thanks.

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs (hunk 2 above)
Comment (akrmn) [reply compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs:45]: yes, that's more of a long term idea, I opened https://github.com/digital-asset/daml/issues/14690 so we don't forget

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs [hunk 5]
```diff
@@ -957,6 +957,14 @@ data InterfaceInstanceMethod = InterfaceInstanceMethod
   }
   deriving (Eq, Data, Generic, NFData, Show)
 
+-- | The interface and template that identify an interface instance.
+-- Currently not part of the AST.
+data InterfaceInstanceKey = InterfaceInstanceKey
```
[the remark this answers is not in the document]
Comment (akrmn) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs:962]: oh, that's a good point, thank you. I'll rename to `InterfaceInstanceHead`

File: compiler/daml-lf-ast/src/DA/Daml/LF/TemplateOrInterface.hs (hunk 4 above)
Comment (akrmn) [reply compiler/daml-lf-ast/src/DA/Daml/LF/TemplateOrInterface.hs:4]: It doesn't _have_ to be in a new file, but in either of those we would get clashes with [`data Template`](https://github.com/digital-asset/daml/blob/main/compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs#L907). Modules using both `Ast` and this module unqualified would still get clashes, since in Haskell we can't namespace the constructors in an `object` like we do in Scala. The best we can do is import this module qualified, 

```haskell
import qualified DA.Daml.LF.TemplateOrInterface as TemplateOrInterface
```

this way we get the same syntax as in Scala, `TemplateOrInterface.Template`

Comment (remyhaemmerle-da) [reply compiler/daml-lf-ast/src/DA/Daml/LF/TemplateOrInterface.hs:4]: fair enough
